---
title: "Data analysis Gregor et al"
output:
  pdf_document: default
  html_notebook: default
---

The following is the code to create the figures in the paper "Mammalian gut metabolomes mirror microbiome composition and host phylogeny"

Load packages
```{r}

library(phyloseq)
library(vegan)
library(ggplot2)
library(dplyr)
library(reshape2)
library(tidyr)
```

Load phyloseq objects and metadata

```{r}
#metadata
metadata <- read.delim("metadata.txt", header = T, sep = "") ## load metadata file
metadata <-subset(metadata, Type=="Feces")
metadata <- droplevels(metadata)


#16S sequencing results
ps_asv <- readRDS("Phyloseq_files/ps_asv.rds")

#LC-MS/MS metabolomics data (fecal samples, filtered to ~10k fecal peaks)
ps_pol <- readRDS("Phyloseq_files/ps_pol.rds")

#GC-MS metabolomics data (fecal samples)
ps_gcms <- readRDS("Phyloseq_files/ps_gcms.rds")

# PICRUSt data
#ps_picrust <- readRDS("Phyloseq_files/ps_picrust.rds")

##Additional files
#LC-MS/MS metabolomics data, all samples and peaks
#ps_polar_all <- readRDS("Phyloseq_files/ps_polar_all.rds")
 
#GC-MS data, all samples
#ps_gcms_all <- readRDS("Phyloseq_files/ps_gcms_all.rds")
```


##Identification of candidate differential peak features


Note: the final results of this code exist in the files in the Indval folder, which can be used to directly generate the heatmap in Fig 2 without running this section. 

Choose a dataset to work with and extract loadings
```{r}

##Choose a dataset: ps_gcms for GCMS or ps_pol for LCMS
ps <- ps_gcms

metab <- as(otu_table(ps), "matrix")

metab_pca <- prcomp(metab)

##extract all loadings as matrix
loadings <- as.data.frame(metab_pca$rotation[,1:2])

##Thresholds: 0.99/0.01 for LCMS, change to 0.95/0.05 for GCMS

PC1_up <- quantile(loadings$PC1, probs = 0.95)
PC1_low <- quantile(loadings$PC1, probs = 0.05)
PC2_up <- quantile(loadings$PC2, probs = 0.95)
PC2_low <- quantile(loadings$PC2, probs = 0.05)


```


Extract a shortlist of loadings that most contribute to PC1 and PC2
```{r}

##set general cutoffs. 
loadingsPC1 <- loadings[loadings[,1]>PC1_up,]
loadingsPC2 <- loadings[loadings[,2]>PC2_up,]

#and to add the neg values 
loadingsPC1neg <- loadings[loadings[,1] < PC1_low,]
loadingsPC2neg <-loadings[loadings[,2] < PC2_low,]

##Create list of unique candidate significant loadings
candidate_loadings <-rbind(loadingsPC1,loadingsPC2,loadingsPC1neg,loadingsPC2neg)
candidate_loadings <- unique(candidate_loadings)

##Visualize the loadings-- all loadings in black, top percent of loadings in red
plot(loadings, col="black")
points(candidate_loadings, col="red")
```


Subset data to list of loadings

```{r}

metab <- metab[,colnames(metab) %in% rownames(candidate_loadings)]

##extract metadata
meta <- as(sample_data(ps), "data.frame")

#match order of samples in metadata and data
meta <- meta[rownames(metab) %in% rownames(meta),]
```

IndVal analysis with indicspecies package
```{r}
library(indicspecies)

##The following is slow, especially for the LCMS data

indval_diet <- multipatt(metab, meta$diet, control = how(nperm=99999))

##look at a summary
summary(indval_diet)

#Bonferroni correction for p values
indval_diet$sign$p.value <- p.adjust(indval_diet$sign$p.value, method="bonferroni")

#Continue with the "sign" matrix
indval_diet <- indval_diet$sign
indval_diet$id <- rownames(indval_diet)

#make the relevant groups more readable
indval_diet <- indval_diet %>% 
  gather(groups, value, -stat,-index,-p.value,-id)%>%  #Change to long format
  filter(value==1)%>%  #Filter for value which are 1
  group_by(id, stat, index, p.value) %>%
  summarise(groups= paste0(groups,collapse=",")) 


```



Repeat for other three groupings. Note, each indval calculation is slow, so this will take a while to generate, especially for the host species 

```{r}
##For the hosts max.order=2 was chosen (ie, combinations of max two host species), for order max is 3
indval_ord <- multipatt(metab, meta$HostOrder, control = how(nperm=99999),max.order = 3)
indval_gut <- multipatt(metab, meta$digestive_strategy, control = how(nperm=99999))
indval_host <- multipatt(metab, meta$HostSpecies, control = how(nperm=99999),max.order = 2)

#Bonferroni correction for p values
indval_gut$sign$p.value <- p.adjust(indval_gut$sign$p.value, method="bonferroni")
indval_ord$sign$p.value <- p.adjust(indval_ord$sign$p.value, method="bonferroni")
indval_host$sign$p.value <- p.adjust(indval_host$sign$p.value, method="bonferroni")

#Continue with the "sign" matrix

indval_gut <- indval_gut$sign
indval_ord <- indval_ord$sign
indval_host <- indval_host$sign

indval_gut$id <- rownames(indval_gut)
indval_ord$id <- rownames(indval_ord)
indval_host$id <- rownames(indval_host)

#make the relevant groups more readable

indval_gut <- indval_gut %>%
  gather(groups, value, -stat,-index,-p.value,-id)%>%  #Change to long format
  filter(value==1)%>%  #Filter for value which are 1
  group_by(id, stat, index, p.value) %>%
  summarise(groups= paste0(groups,collapse=","))

indval_host <- indval_host %>%
  gather(groups, value, -stat,-index,-p.value,-id)%>%  #Change to long format
  filter(value==1)%>%  #Filter for value which are 1
  group_by(id, stat, index, p.value) %>%
  summarise(groups= paste0(groups,collapse=","))

indval_ord <- indval_ord %>%
  gather(groups, value, -stat,-index,-p.value,-id)%>%  #Change to long format
  filter(value==1)%>%  #Filter for value which are 1
  group_by(id, stat, index, p.value) %>%
  summarise(groups= paste0(groups,collapse=","))

```

Rename the p value columns before merging the dataframes 
```{r}
colnames(indval_diet)[colnames(indval_diet) == 'p.value'] <- 'p.value_diet'
colnames(indval_gut)[colnames(indval_gut) == 'p.value'] <- 'p.value_gut'
colnames(indval_ord)[colnames(indval_ord) == 'p.value'] <- 'p.value_ord'
colnames(indval_host)[colnames(indval_host) == 'p.value'] <- 'p.value_host'
```



To make Table S2, merge all the results into one table
```{r}
indval <- merge(indval_gut, indval_diet, by="id")
indval <- merge(indval, indval_ord, by="id")
indval <- merge(indval, indval_host, by="id")

rownames(indval) <- indval$id
```


For GCMS results: Add the metadata for GCMS
```{r}
#load the lists of analogues with ID
##GC-MS
GCMS <- read.delim("Indval/GCMS-GNPS_ID_file.txt", header = TRUE, check.names=F, row.names=1)

##now get the cluster ID isolated to match to ID file
split <- read.table(text = rownames(indval), sep = "_", colClasses = "character")

##combine back with list of loadings
indval_IDs <- cbind(indval, split)

GCMS_subset <- GCMS[match(indval_IDs$V1,GCMS$name),] #gcms

##for GC-MS, combine with indval
GCMS_indval <- cbind(GCMS_subset,indval_IDs)
rownames(GCMS_indval) <- rownames(indval_IDs)

#Remove metabolites which are not significant for any grouping

pvals <- subset(GCMS_indval, select= c("p.value_gut", "p.value_diet", "p.value_ord", "p.value_host"))

##convert NA vals to 1
pvals[is.na(pvals)] <- 1

##Make a list of the non-sig ones
pvals_nonsig <- filter_all(pvals, all_vars(. > 0.05))

##Make a list of the significant ones
sig <- setdiff(rownames(pvals), rownames(pvals_nonsig))

#Keep only significant ones

GCMS_indval_sig <- GCMS_indval[rownames(GCMS_indval) %in% sig,]
```


For LC-MS results: Add the metadata for LCMS
```{r}
#load the lists of analogues with ID

##LC-MS
LCMS <- read.delim("Indval/LCMS-GNPS_ID_file.txt", header = TRUE, check.names=F, row.names=1)
##trim out irrelevant columns
LCMS <- LCMS[,37:51]


##now get the cluster ID isolated to match to ID file
split <- read.table(text = rownames(indval), sep = "_", colClasses = "character")

##combine back with list of loadings
indval_IDs <- cbind(indval, split)


##match them and make the rownames into the full metabolite names
LCMS_subset <- LCMS[match(indval_IDs$V1,rownames(LCMS)),] #lcms

rownames(LCMS_subset) <- rownames(indval_IDs)

##For all LCMS metabolites, combine with indval 
LCMS_indval <- cbind(LCMS_subset,indval_IDs)
rownames(LCMS_indval) <- rownames(indval_IDs)


#Remove metabolites which are not significant for any grouping

pvals <- subset(LCMS_indval, select= c("p.value_gut", "p.value_diet", "p.value_ord", "p.value_host"))

##convert NA vals to 1
pvals[is.na(pvals)] <- 1

##Make a list of the non-sig ones
pvals_nonsig <- filter_all(pvals, all_vars(. > 0.05))

##Make a list of the significant ones
sig <- setdiff(rownames(pvals), rownames(pvals_nonsig))

#Keep only significant ones

LCMS_indval_sig <- LCMS_indval[rownames(LCMS_indval) %in% sig,]



```

For the LCMS results, all of these metabolites were manually curated based on the annotations here, the networks, the MS2 spectra, and the NAP and MolNetEnhancer results, resulting in the list of annotated metabolites in Table S2.


##Figure 2B- Heatmap 

Make a vector to arrange the samples by host species/order
```{r}
animals <- c("Hyena_1","Jungle_cat_1","Jungle_cat_2","Sand_cat_1","Wolf_1","Wolf_2","Leopard_1","Leopard_2","Lion_1","Lion_2","Lion_3","Tiger_1","Tiger_2","Af_elephant_1","Af_elephant_2","Af_elephant_3","As_elephant_1","As_elephant_2","As_elephant_3","As_elephant_4","As_elephant_5","Donkey_1","Zebra_1","Zebra_10","Zebra_11","Zebra_12","Zebra_13","Zebra_14","Zebra_15","Zebra_16","Zebra_2","Zebra_3","Zebra_4","Zebra_5","Zebra_6","Zebra_7","Zebra_8","Zebra_9","Rhino_1","Rhino_10","Rhino_11","Rhino_12","Rhino_13","Rhino_2","Rhino_3","Rhino_4","Rhino_5","Rhino_6","Rhino_7","Rhino_8","Rhino_9","Goat_1","Sheep_1","Sheep_2","Sheep_3","Sheep_4","Sheep_5","Black_bear_1","Black_bear_2","Brown_bear_1","Brown_bear_2","Brown_bear_3","Brown_bear_4","Brown_bear_5","Coati_1","Coati_3","Coati_4","Coati_5","Coati_6","Coati_7","Coati_8","Chimpanzee_1","Chimpanzee_10","Chimpanzee_11","Chimpanzee_12","Chimpanzee_13","Chimpanzee_4","Chimpanzee_5","Chimpanzee_6","Chimpanzee_7","Chimpanzee_9","Gorilla_1","Gorilla_2","Gorilla_3","Gorilla_4","Gorilla_5","Gorilla_6","Mandarill_1","Mandarill_2","Mandarill_3","Mandarill_4","Mandarill_5","Mandarill_6","Mandarill_7","Mandarill_8","Capuchin_1","Gibbon_1","Gibbon_2","Gibbon_3","Lemur_1","Lemur_2")
```

Fig 2B: This is a heatmap of significantly enriched metabolites in the LC-MS/MS data which could be classified.
```{r}

ps <- ps_pol

#Load annotated peaks
annotations <-read.csv("Indval/LCMS-indval_sig_with_IDs.csv", header = TRUE, check.names=F, row.names = 1)

##Prune phyloseq object down to the 76 peaks
ps_sig <- prune_taxa(rownames(annotations), ps)

##Add annotations as a taxonomy table
tax_table(ps_sig) <- as.matrix(annotations)

ps_sig

```

Make a heatmap ordered by samples in above vector and chemical family (Fig 2B)
```{r}

p <- plot_heatmap(ps_sig, method= NULL, distance= NULL,  sample.label = "HostSpecies", low="#66CCFF", high="#000033", na.value="white",sample.order = animals, taxa.label = "Name", taxa.order = rownames(annotations)) 

p
```

Additional heatmaps

Fig S12: unidentified LC-MS/MS differential peak features
```{r}

#Load all significant peaks
all_sig <-read.csv("Indval/LCMS-indval_sig_all.csv", header = TRUE, check.names=F, row.names = 1)

no_ID <- setdiff(rownames(all_sig), rownames(annotations))

ps_sig_noID <- prune_taxa(no_ID, ps_pol)


p <- plot_heatmap(ps_sig_noID, method= "NMDS", distance= "bray",  sample.label = "HostSpecies", low="#66CCFF", high="#000033", na.value="white",sample.order = animals) 

p
```
Fig S13: Dietary triterpenoids

```{r}

##The following nodes were found in the GNPS network analysis in the triterpenoid cluster
diettrp <- c(
"11315_431.314_4.485", #oat straw
"11317_431.314_4.887",  #oat straw
"2293_445.293_5.172", #wheat straw
"13556_461.287_4.8",#wheat straw
"338_449.324_6.143",#pellets
"412_449.324_6.268"#pellets
)

##These peaks appear in dietary samples, so are in the unfiltered LC-MS dataset

ps_polar_all <- readRDS("Phyloseq_files/ps_polar_all.rds")

diettrp <- rev(diettrp)

ps_diettrp <- prune_taxa(diettrp, ps_polar_all)


p <- plot_heatmap(ps_diettrp, method= NULL, distance= NULL, 
                  sample.label = "HostSpecies", sample.order = animals,
                  low="#66CCFF", high="#000033", na.value="white",  
                  taxa.order=rev(diettrp)) 

p

```



Fig S14: GC-MS differential peak features

```{r}
#Load GCMS indval results
gcms_sig <-read.csv("Indval/GCMS-indval_sig.csv", header = TRUE, check.names=F, row.names = 1)

ps_sig_GCMS <- prune_taxa(rownames(gcms_sig), ps_gcms)

tax_table(ps_sig_GCMS) <- gcms_sig


p <- plot_heatmap(ps_sig_GCMS, method= "NMDS", distance= "bray",  
                  sample.label = "HostSpecies", 
                  low="#66CCFF", high="#000033", na.value="white",
                  sample.order = animals, taxa.order = rev(rownames(gcms_sig))) 

p
```

